BDArmory/Weapons/ModuleWeapon.cs

```diff
@@ -4554,7 +4555,7 @@ void CheckFinalFire()
                         roundsPerMinute = Mathf.Lerp(baseRPM, (baseRPM / 10), spooltime);
                     }
                 }
-                if (ChargeTime > 0) hasCharged = false;
+                if (ChargeTime > 0 && hasCharged && timeSinceFired > 1) hasCharged = false;
```

Comment:
Yes. Was originally GetTimeGap() * 2 in the original draft, but that would result in very short intervals for rapidfire weapons, using chargeTime (maybe with a multiplier?) is probably the most sensible option

Reply:
I think that it'd probably depend on the weapon. E.g., a massive railgun could have a large charge time and be fully discharged in a fraction of a second, whereas a multi-shot pulse laser could have a capacitor that allows for a burst of 10-20 shots before needing recharging.

Reply:
``hasCharged`` is only set for weapons that only need to do an initial charge to prime the weapon before they fire continuously (e.g. a gatling gun needing to fully spin up before it fires); stuff that has to charge before each shot (i.e. railguns) wouldn't be affected by this. How many shots a pulse laser or similar gets before it needs to charge again would be controlled elsewhere, either heatPerShot/roundsPerMag.
Though you raise a good point about capacitors - those would hold on to whatever is left in them if firing prematurely stops, presumably (unless switching away from the weapon discharges them as a safety measure or something?)
We could add a new field for ``chargeHoldLength`` that would allow setting how long the trigger can be released before the charge needs to be redone.

Reply:
I thought the spin-up time (at least for the gatling gun example) was handled by `SpoolUpTime`? Or is that for weapons that just fire more slowly while spinning up vs needing to fully charge before beginning to fire?

For the `timeSinceFired > ChargeTime * 2` check, can a weapon have a RoF high enough such that the `timeGap` is less than the `ChargeTime`? If so, then `ChargeTime` should be clamped to at least the `timeGap`.
I like the idea of having a `chargeHoldLength` better though. I think something like the following should work:
```diff
diff --git a/BDArmory/Weapons/ModuleWeapon.cs b/BDArmory/Weapons/ModuleWeapon.cs
index 661173de0..6e2447a2d 100644
--- a/BDArmory/Weapons/ModuleWeapon.cs
+++ b/BDArmory/Weapons/ModuleWeapon.cs
@@ -1842,2 +1842,3 @@ void Update()
 
+        float chargeHoldTimer = 0, chargeHoldLength = 5; // FIXME Just for demo purposes.
         void FixedUpdate()
@@ -1862,3 +1863,13 @@ void FixedUpdate()
                 }
-                else if (eWeaponType == WeaponTypes.Laser)
+                else
+                {
+                    if (ChargeTime > 0 && hasCharged)
+                    {
+                        if ((chargeHoldTimer -= Time.fixedDeltaTime) > 0)
+                        {
+                            chargeHoldTimer = 0;
+                            hasCharged = false;
+                        }
+                    }
+                    if (eWeaponType == WeaponTypes.Laser)
                     {
@@ -1870,2 +1881,3 @@ void FixedUpdate()
                     }
+                }
                 vessel.GetConnectedResourceTotals(AmmoID, out double ammoCurrent, out double ammoMax); //ammo count was originally updating only for active vessel, while reload can be called by any loaded vessel, and needs current ammo count
@@ -4566,3 +4578,3 @@ void CheckFinalFire()
                 }
-                if (ChargeTime > 0 && hasCharged && timeSinceFired > ChargeTime * 2) hasCharged = false;
+                // if (ChargeTime > 0 && hasCharged && timeSinceFired > ChargeTime * 2) hasCharged = false;
             }
@@ -5458,3 +5470,3 @@ IEnumerator ChargeRoutine()
             {
-                chargeState.normalizedTime = 0;
+                chargeState.normalizedTime = chargeHoldTimer / chargeHoldLength;
                 chargeState.enabled = true;
@@ -5468,3 +5480,3 @@ IEnumerator ChargeRoutine()
             {
-                yield return new WaitForSecondsFixed(ChargeTime);
+                yield return new WaitForSecondsFixed(ChargeTime * (1 - chargeHoldTimer / chargeHoldLength));
             }
@@ -5472,3 +5484,7 @@ IEnumerator ChargeRoutine()
             isCharging = false;
-            if (!ChargeEachShot) hasCharged = true;
+            if (!ChargeEachShot)
+            {
+                hasCharged = true;
+                chargeHoldTimer = chargeHoldLength;
+            }
             switch (eWeaponType)
```

Separately, I notice that there's a `ChargeEachShot` field too, which prevents `hasCharged` from ever being `true` so those weapons will never reach the code block that actually fires the weapon.

Reply:
> I thought the spin-up time (at least for the gatling gun example) was handled by SpoolUpTime? Or is that for weapons that just fire more slowly while spinning up vs needing to fully charge before beginning to fire?  

Correct. ``SpoolUpTime`` is for spin up where the weapon can fire immediately with reduced RoF, and gains RoF as it spins up to full.
> For the timeSinceFired > ChargeTime * 2 check, can a weapon have a RoF high enough such that the timeGap is less than the ChargeTime?

Quite easily, if you have a weapon with, say, a half-second charge time that fired 600 RPM, the timeGap would only be .1s. In which case the check would return false and ``hasCharged`` would remain true, and you could release the trigger for up to a second and then resume firing without having to charge again.
 Did you mean if the timeGap was more than the chargeTime (i.e. .5s chargeTime, with 60 RPM/1s timegap)?
 > Separately, I notice that there's a ChargeEachShot.
 
 Yes. ``chargeEachShot`` is what determines if hasCharged is used or not.
``ChargeEachShot`` triggers the ChargeRoutine coroutine; ``hasCharged`` is set within the coroutine; if ``ChargeEachShot`` is false, then it uses ``hasCharged``, else ``chargeEachShot`` supersedes ``hasCharged``, correct.
Weapon charge is set by having a ``chargeTime`` > 0; if it is, then there are two types of charge a weapon can do, set by the T/F state of ``chargeEachShot`` in the .cfg;, charging before each shot (e.g. the BDA-E railgun), or charging once then firing full auto until the weapon overheats/reloads/trigger is released

Reply:
Ah, I see. So the charge routine charges the weapon and then fires it once it's charged and if `chargeEachShot` is `false`, then it can continue firing in subsequent frames if `finalFire` is `true`. That makes sense to me now.
I think the timing with the `yield WaitForFixedUpdate` and `AimAndFire` means that it maintains that order of operations so that weapons with `chargeEachShot = false` don't fire twice on the same frame, but that'd be an edge case I can check later.

With the `chargeHoldTimer`, I was thinking that it was separate from `timeSinceFired` where it would immediately start to drain once it was charged, but that doesn't make much sense now that I think about it again.
One problem with using `timeSinceFired` is that if it charges but doesn't have a firing solution for a while, then it'll discharge. Is that actually how we want it or is it better that it remains charged until `AimAndFire` is called without `aimAndFireIfPossible` being `true`?
If it's the latter, then it might be better to immediately set `hasCharged = false` and run the `chargeHoldTimer` at the start of `AimAndFire` if `aimAndFireIfPossible` is `false`, and then when `aimAndFireIfPossible` becomes `true` again, you'd restart the `ChargeRoutine` from a partially charged state.

Reply:
> Is that actually how we want it or is it better that it remains charged until AimAndFire is called without aimAndFireIfPossible being true?

That would make it similar to how weapons with spool up currently work, if it starts firing and then the target moves out of position and ``finalFire`` becomes false. Assuming the ``chargeHoldLength`` is set to something short. Setting it to longer would make so the weapon would remain primed until ``finalFire`` becomes true again.

Reply:
It already does both.
Every time the weapon fires, ``timeSinceFired`` updates because ``timeSinceFired => Time.time - timeFired``;  as long as the trigger is held and the weapon is firing, ``hasCharged`` will remain true. 
If the trigger is released, then it will time out if it isn't refired within the time limit. 
If the trigger is held, the weapon fires. The only time holding the trigger wouldn't fire the weapon is if it's overheated/reloading/pointing at self, and the first two already reset ``hasCharged``.

Reply:
Ah, OK. I see I'm reading wrongly when the trigger is being held, which is only when `finalFire` is `true` (I can only blame a lack of sleep).
This should be good to merge then.